File: symmetricString.py
def split(word):
    # split string into list of chars
    # https://www.geeksforgeeks.org/python-split-string-into-list-of-characters/
    return [char for char in word]


symmetricLetters = ['A', 'H', 'I', 'O', 'W', 'V', 'M', 'T', 'U', 'X']
symmetricPairs = [('9', 'P'), ('A', 'A'), ('O', 'O'), ('J', 'L')]

failed = False

inputString = "NATTAN"
inputStringList = split(inputString)

if len(inputString) % 2 == 0:
    # if even
    for i in range(int(len(inputString)/2)):
        # go through each letter of string until it gets to half point
        if not failed:
            pairFound = False
            for j in range(len(symmetricPairs)):
                # go through each symmetric pair and try to find a match
                firstLetter = inputStringList[i]
                lastLetter = inputStringList[len(inputString) - 1 - i]
                pair1 = symmetricPairs[j][0]
                pair2 = symmetricPairs[j][1]

                if (firstLetter == pair1 and lastLetter == pair2) or (lastLetter == pair1 and firstLetter == pair2):
                    pairFound = True
                    break
            if not pairFound:
                failed = True
                break
else:
    # if odd: pretty much use the even algorithm, but just check if the single letter in the middle is symmetric
    for i in range(int(len(inputString)/2)):
        # go through each letter of string until it gets to half point
        if not failed:
            pairFound = False
            for j in range(len(symmetricPairs)):
                # go through each symmetric pair and try to find a match
                firstLetter = inputStringList[i]
                lastLetter = inputStringList[len(inputString) - 1 - i]
                pair1 = symmetricPairs[j][0]
                pair2 = symmetricPairs[j][1]

                if (firstLetter == pair1 and lastLetter == pair2) or (lastLetter == pair1 and firstLetter == pair2):
                    pairFound = True
                    break
            if not pairFound:
                failed = True
                break

    middleLetterIsSymmetric = False
    for i in range(len(symmetricLetters)):
        # find if middle letter is a symmetric letter. middle letter of an odd length string can be found by using the "div" operator. cse15 y'all already know
        if inputStringList[len(inputString) // 2] == symmetricLetters[i]:
            middleLetterIsSymmetric = True
            break
    if not middleLetterIsSymmetric:
        failed = True

if not failed:
    print("true")
else:
    print("false")

File: symmetricString.cpp
#include <iostream>
#include <string>
bool symmetric(std::string s)
{
    char symmetricLetters[] = {
        'A', 'H', 'I', 'M', 'O', 'T', 'U', 'V', 'W', 'X', 'Y', '8', '0'};

    std::pair<char, char> symmetricPairs[] = {{'9', 'P'}, {'A', 'A'}, {'I', 'I'}, {}};

    bool failed = false;
    bool pairFound;
    bool middleSymmetry;
    char firstLetter;
    char lastLetter;
    char pair1, pair2;
    std::string s = "NATE";

    if (s.size() % 2 == 0)
    {
        for (int i = 0; i < s.size() / 2; i++)
        {
            if (failed == false)
            {
                pairFound = false;
                for (int j = 0; j < s.size() / 2; j++)
                {
                    firstLetter = s[i];
                    lastLetter = s[s.size() - i - 1];
                    pair1 = symmetricPairs[j].first;
                    pair2 = symmetricPairs[j].second;

                    if ((firstLetter == pair1) && (lastLetter == pair2) || (firstLetter == pair2) && (lastLetter == pair1))
                    {
                        pairFound = true;
                        break;
                    }
                }
                if (pairFound = false)
                {
                    failed = true;
                    break;
                }
            }
        }
    }
    else
    {
        for (int i = 0; i < s.size() / 2; i++)
        {
            if (failed == false)
            {
                pairFound = false;
                for (int j = 0; j < s.size() / 2; j++)
                {
                    firstLetter = s[i];
                    lastLetter = s[s.size() - i - 1];
                    pair1 = symmetricPairs[j].first;
                    pair2 = symmetricPairs[j].second;

                    if ((firstLetter == pair1) && (lastLetter == pair2) || (firstLetter == pair2) && (lastLetter == pair1))
                    {
                        if ((j == i) && (j || i == symmetricLetters[]))
                        {
                            pairFound = true;
                            break;
                        }
                    }
                }
                if (pairFound = false)
                {
                    failed = true;
                    break;
                }
            }
        }
    }
}